contrib: check if the user has cargo-public-api

storopoli opened the pull request:
Closes #3764.

yancyribbens can you test it locally? It works in my end both with and without `cargo-public-api`. Of course without gives the error message that I was supposed to see.

EDIT: You can run it with `just check-api`.

tcharding reviewed (dismissed):
ACK c4913c14d360b08f0dd60ce7365fc2e5c4e83bdc

tcharding commented:
Cheers

storopoli commented:
NOTE: CI is broken probably because of #3758 because

> Some wombat queued 100 PRs up and is constantly breaking the api text files.

😂

yancyribbens commented:
I don't have just installed because I can just type `./contrib/check-for-api-changes.sh`.  Can confirm it works now though.  Thanks!

storopoli commented on contrib/check-for-api-changes.sh:
```diff
@@ -81,6 +82,13 @@ need_nightly() {
+    if command -v cargo-public-api > /dev/null; then
+        return
+    fi
+    err "cargo-public-api is not installed; please run 'cargo install cargo-public-api' or 'cargo binstall cargo-public-api'"
```
Note: do we add `+nightly` as @yancyribbens reports [here](https://github.com/rust-bitcoin/rust-bitcoin/issues/3764#issuecomment-2542451548) so this becomes:

```suggestion
    err "cargo-public-api is not installed; please run 'cargo +nightly install --locked cargo-public-api'"
```

Yeah I think this is fair. Will do the changes...

yancyribbens replied:
Sorry I don't think `+nightly` was actually required.  Not sure it really matters either way.

storopoli replied:
Oh ok!

tcharding commented:
You are changing the first patch in the second one @storopoli, can you move that change into the first patch please.

Also, why not just use the same install command as whats in `api/README.md`?

```
cargo +stable install cargo-public-api --locked
```

storopoli commented:
Done!

tcharding reviewed (approved):
ACK 6fabf52a7df35d70b259d86fa5dfbaf4d1e78f9d

tcharding commented:
Thanks man.

apoelstra reviewed (dismissed):
ACK 78ed973f55c7a02f2150ab46d02850103c120f7a; successfully ran local tests

apoelstra commented:
In 6fabf52a7df35d70b259d86fa5dfbaf4d1e78f9d:

This says `+stable` but actually the user needs to install it with the `nightly-version` compiler.

storopoli commented:
> In [6fabf52](https://github.com/rust-bitcoin/rust-bitcoin/commit/6fabf52a7df35d70b259d86fa5dfbaf4d1e78f9d):
> 
> This says `+stable` but actually the user needs to install it with the `nightly-version` compiler.

Ok I've changed it to `+nightly` and also updated the instructions in `api/README.md` in f01f071751332bb6aa8b9affa4677f913035de7f

apoelstra reviewed (approved):
ACK f01f071751332bb6aa8b9affa4677f913035de7f; successfully ran local tests

tcharding commented:
Ugh, sorry. Perhaps I messed up when I initially wrote the docs. I did not verify it though.

tcharding commented:
My previous post didn't sit well with me, went back and checked. I am able to install using `cargo +stable install cargo-public-api --locked` as I originally wrote in the docs. I am also able to then run `just check-api`.

apoelstra commented:
Our script right now checks that we have a nightly compiler before running. Should we change it to use stable?

tcharding commented:
We need nightly to run it but we don't need nightly to install it.

apoelstra commented:
@tcharding if you approve this I'll go ahead and merge it. I'm not sure what the best course of action here is. I *suspect* the most correct thing is to tell the user to install it with `nightly-version` because this ensures that they *have* nightly-version installed and hints that they should use it.

But probably anything is fine. I don't expect many people to actually use this. I think most drive-by contributors will get a red X and then we'll need to advise what to do about the API break anyway, and one of us can just push a fix commit if we think it's ok.

tcharding reviewed (approved):
ACK f01f071751332bb6aa8b9affa4677f913035de7f

tcharding commented:
I removed the mention from the PR description (so merge script doesn't spam).
